### PR TITLE
Fix SDN registry startup to not depend on master having been started first

### PR DESCRIPTION
**Pull request by danwinship**
Origin side of https://github.com/openshift/openshift-sdn/pull/248
(Replacing #6682 which was wrong.)

@eparis @smarterclayton

### CODE REVIEW

**Comment by smarterclayton** [comment]
Approved, LGTM [merge]

**Comment by smarterclayton** [comment]
Legitimate bug in the PR

**Comment by dcbw** [comment]
@smarterclayton yeah, danw will be back soon but I'm digging into it.

**Comment by dcbw** [comment]
@smarterclayton @danwinship it looks like a race between master and node.

Note the logs never get print out "Started Origin Controllers" which indicates that the master is fully set up.  startControllers() gets run async and that's where the SDN gets started, and that's where ClusterNetwork("default") gets created if it doesn't yet exist.

But the test code that starts the server only verifies that it can call some API and then lets client startup proceed; it doesn't actually wait for startControllers() to return because that's done from a goroutine.

So the test code proceeds to start the node while the master is still in startControllers().  Node startup has less work to do, so by the time it gets to config.RunSDN() it's raced the master and fails hard on getting ClusterNetwork("default").

Honestly I'd say it's a bug in the testcases, but clearly something that needs to get fixed.

**Comment by smarterclayton** [comment]
I'd recommend doing a retry loop on cluster network using the retry library
- there are other reasons it could fail and a few checks never hurt anyone.
 pkg/util/wait.ExponentialBackoff with a duration of a second and 3-4 tries
(doubling each time, maybe)

**Comment by eparis** [comment]
this PR is being un-proposed for 3.1.1. We will continue to work to resolve the issue, but the risk/reward does not appear to be sufficient to race this in at the last moment.

**Comment by danwinship** [comment]
(Note that the openshift-sdn side hasn't actually been committed there yet, I just pushed it here to get the tests to run before committing it there.)

**Comment by danwinship** [comment]
Latest push pulls in https://github.com/openshift/openshift-sdn/pull/250. There's no loop/exponential backoff, because with this rearragement, the endpoints-related `ClusterNetwork.Get("default")` occurs after another part of the registry code has already verified that it is non-nil.

File: Godeps/_workspace/src/github.com/openshift/openshift-sdn/plugins/osdn/registry.go [hunk 1]
```diff
@@ -611,6 +602,14 @@ func getEvent(eventQueue *oscache.EventQueue, startVersion uint64, checkConditio
 // FilteringEndpointsConfigHandler implementation
 func (registry *Registry) SetBaseEndpointsHandler(base pconfig.EndpointsConfigHandler) {
 	registry.baseEndpointsHandler = base
+
+	cn, err := registry.oClient.ClusterNetwork().Get("default")
+	if err != nil {
+		// "can't happen"; StartNode() will already have ensured that there's no error
+		panic("Failed to get ClusterNetwork: " + err.Error())
```
Comment (liggitt) [inline Godeps/_workspace/src/github.com/openshift/openshift-sdn/plugins/osdn/registry.go]: Why not log.Fatalf like other places in this file?

Comment (danwinship) [inline Godeps/_workspace/src/github.com/openshift/openshift-sdn/plugins/osdn/registry.go]: fixed

**Comment by danwinship** [comment]
[test]

**Comment by danwinship** [comment]
[test]

**Comment by danwinship** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Noticing here that the plugin is coupled to the service proxy - when we move the service proxy and the plugins out of node.go, how is this code going to work?

**Comment by danwinship** [comment]
> when we move the service proxy and the plugins out of node.go, how is this code going to work?

I don't know... I hadn't heard about that. Is there a PR/issue for it?

(The coupling is to keep people from being able to break isolation by manually creating service endpoints pointing to another tenant's pods. The source network ID gets lost when the packet moves from OVS to iptables, so we have to decide whether the packet is allowed through before that point. Maybe we could just make it impossible to manually create endpoints pointing into ClusterNetworkCIDR or ServicesNetworkCIDR?)

**Comment by smarterclayton** [comment]
Maybe.  Can you add a card / TODO in openshift-sdn to track that?  We might have to have a way for openshift-sdn to ask the kube-proxy for info about endpoints or similar.

**Comment by danwinship** [comment]
https://trello.com/c/80S5GwWW

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**